File: Engine/GameEngine/GameObject.cpp
//
//  GameObject.cpp
//  LetsUseSwift
//
//  Created by Sora Sugiyama on 3/5/26.
//

#include "GameObject.hpp"
#include <Metal/Metal.hpp>

ObjectList::ObjectList():
size{0U}
{}

ObjectList::~ObjectList() = default;

uint8_t ObjectList::add_back(void* pVertexBuffer, void *pIndexBuffer, std::uint32_t n, bool isUInt32, float y0)
{
    if (size >= MAX_SIZE) {
        std::cerr << "Out of Bounds" << std::endl;
        return 0U;
    }
    
    GameObject &back = arr[size++];
    
    back.vertexBuffer = NS::RetainPtr(static_cast<MTL::Buffer*>(pVertexBuffer));
    back.indexBuffer = NS::RetainPtr(static_cast<MTL::Buffer*>(pIndexBuffer));
    
    back.indexCount = n;
    back.isUInt32Index = isUInt32;
    back.y0 = y0;
    return 1U;
}

void ObjectList::pop(const std::size_t index)
{
    if (index >= size || index <0) {
        return;
    }
    
    --size;
    GameObject &cur = arr[index], &back = arr[size];
    
    if(size != index) {
        cur = std::move(back);
    }
    
    back = GameObject();
}

const GameObject& ObjectList::operator[](const std::size_t index) const
{
    return arr[index];
}

File: SimpleEngine/Engine/GameObject.swift
//
//  GameObject.swift
//  SimpleEngine
//
//  Swift port of the C++ GameObject / ObjectList
//

import Metal

struct GameObject {
    var vertexBuffer: MTLBuffer?
    var indexBuffer:  MTLBuffer?
    var indexCount:   UInt32 = 0
    var isUInt32Index: Bool  = false
    var y0: Float = 0
}

final class ObjectList {
    private static let maxSize = 1024

    private var arr: ContiguousArray<GameObject>
    private(set) var count: Int = 0

    init() {
        // ContiguousArray gives contiguous memory layout for better cache performance
        arr = ContiguousArray(repeating: GameObject(), count: ObjectList.maxSize)
    }

    @discardableResult
    func addBack(vertexBuffer: MTLBuffer, indexBuffer: MTLBuffer,
                 indexCount: UInt32, isUInt32: Bool, y0: Float = 0) -> Bool {
        guard count < ObjectList.maxSize else {
            print("ObjectList: Out of Bounds")
            return false
        }
        arr[count] = GameObject(
            vertexBuffer: vertexBuffer,
            indexBuffer:  indexBuffer,
            indexCount:   indexCount,
            isUInt32Index: isUInt32,
            y0: y0
        )
        count += 1
        return true
    }

    func pop(at index: Int) {
        guard index >= 0 && index < count else { return }
        count -= 1
        if index != count {
            arr[index] = arr[count]
        }
        arr[count] = GameObject()
    }

    subscript(index: Int) -> GameObject { arr[index] }
}
